Sources/SwiftSyntax/SyntaxParser.swift

```diff
@@ -135,9 +144,81 @@ public enum SyntaxParser {
       swiftparse_parser_set_node_lookup(c_parser, nodeLookup);
     }
 
+    // Set up diagnostics consumer if requested by the caller.
+    if let diagConsumer = diagConsumer {
+      // If requested, we should set up a source location converter to calculate
+      // line and column.
+      let converter = diagConsumer.calculateLineColumn ?
+        SourceLocationConverter(file: file, source: source) : nil
+      let diagHandler = { (diag: CDiagnostic!) in
+        diagConsumer.handle(Diagnostic(diag: diag, using: converter))
+      }
+      swiftparse_parser_set_diagnostic_handler(c_parser, diagHandler)
+    }
+
     let c_top = swiftparse_parse_string(c_parser, source)
 
   // Get ownership back from the C parser.
     return RawSyntax.moveFromOpaque(c_top)!
   }
 }
+
+extension SourceRange {
+  init(_ range: CRange, _ converter: SourceLocationConverter?) {
+    let start = SourceLocation(offset: Int(range.offset), converter: converter)
+    let end = SourceLocation(offset: Int(range.offset + range.length),
+                             converter: converter)
+    self.init(start: start, end: end)
+  }
+}
+
+extension Diagnostic.Message {
+  init(_ diag: CDiagnostic) {
+    let message = String(cString: swiftparse_diagnostic_get_message(diag))
+    switch swiftparse_diagnostic_get_severity(diag) {
+    case SWIFTPARSER_DIAGNOSTIC_SEVERITY_ERROR:
+      self.init(.error, message)
+    case SWIFTPARSER_DIAGNOSTIC_SEVERITY_WARNING:
+      self.init(.warning, message)
+    case SWIFTPARSER_DIAGNOSTIC_SEVERITY_NOTE:
+      self.init(.note, message)
+    default:
+      fatalError("unrecognized diagnostic level")
+    }
+  }
+}
+
+extension FixIt {
+  init(_ cfixit: CFixit, _ converter: SourceLocationConverter?) {
+    let replacement = String(cString: cfixit.text)
+    let range = SourceRange(cfixit.range, converter)
+    if cfixit.range.length == 0 {
+      // Insert
+      self = .insert(SourceLocation(offset: Int(cfixit.range.offset),
+                                    converter: converter), replacement)
+    } else if replacement.isEmpty {
+      // Remove
+      self = .remove(range)
+    } else {
+      // Replace
+      self = .replace(range, replacement)
+    }
+  }
+}
+
+extension Diagnostic {
+  init(diag: CDiagnostic, using converter: SourceLocationConverter?) {
+    // Collect highlighted ranges
+    let hightlights = (0..<swiftparse_diagnostic_get_range_count(diag)).map {
+      return SourceRange(swiftparse_diagnostic_get_range(diag, $0), converter)
+    }
+    // Collect fixits
+    let fixits = (0..<swiftparse_diagnostic_get_fixit_count(diag)).map {
+      return FixIt(swiftparse_diagnostic_get_fixit(diag, $0), converter)
+    }
+    self.init(message: Diagnostic.Message(diag),
+      location: SourceLocation(offset: Int(swiftparse_diagnostic_get_source_loc(diag)),
+                               converter: converter),
+      notes: [], highlights: hightlights, fixIts: fixits)
```

Comment:
It looks like notes are supposed to be attached to a diagnostic (from `notes: []`). I think what we need to do is keep the last error or warning around in order to attach notes to it.

Reply:
Done in a separate commit.